File: src/nameserver/standalone_test.cc
/*
 * Copyright 2021 4Paradigm
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include <brpc/server.h>
#include <gflags/gflags.h>
#include <sched.h>
#include <unistd.h>

#include "base/file_util.h"
#include "base/glog_wrapper.h"
#include "client/ns_client.h"
#include "gtest/gtest.h"
#include "nameserver/name_server_impl.h"
#include "proto/name_server.pb.h"
#include "proto/tablet.pb.h"
#include "proto/type.pb.h"
#include "rpc/rpc_client.h"
#include "tablet/tablet_impl.h"
#include "test/util.h"

DECLARE_string(endpoint);
DECLARE_string(db_root_path);
DECLARE_string(ssd_root_path);
DECLARE_string(hdd_root_path);
DECLARE_string(zk_cluster);
DECLARE_string(zk_root_path);
DECLARE_int32(zk_session_timeout);
DECLARE_int32(request_timeout_ms);
DECLARE_int32(zk_keep_alive_check_interval);
DECLARE_int32(make_snapshot_threshold_offset);
DECLARE_uint32(name_server_task_max_concurrency);
DECLARE_bool(auto_failover);

using brpc::Server;
using openmldb::tablet::TabletImpl;
using ::openmldb::zk::ZkClient;
using std::shared_ptr;
using std::string;
using std::tuple;
using std::vector;

namespace openmldb {
namespace nameserver {

class MockClosure : public ::google::protobuf::Closure {
 public:
    MockClosure() {}
    ~MockClosure() {}
    void Run() {}
};
class StandaloneTest : public ::testing::Test {
 public:
    StandaloneTest() {}
    ~StandaloneTest() {}
};

TEST_F(StandaloneTest, smoketest) {
    FLAGS_zk_root_path = "/rtidb3" + ::openmldb::test::GenRand();
    brpc::Server tablet;
    ASSERT_TRUE(::openmldb::test::StartTablet("127.0.0.1:9530", &tablet));

    brpc::Server server;
    ASSERT_TRUE(::openmldb::test::StartNS("127.0.0.1:9631", "127.0.0.1:9530", &server));
    ::openmldb::client::NsClient client("127.0.0.1:9631", "");
    ASSERT_EQ(client.Init(), 0);

    TableInfo table_info;
    std::string name = "test" + ::openmldb::test::GenRand();
    table_info.set_name(name);
    ::openmldb::test::AddDefaultSchema(0, 0, ::openmldb::type::kAbsoluteTime, &table_info);
    std::string msg;
    ASSERT_TRUE(client.CreateTable(table_info, false, msg));
}

TEST_F(StandaloneTest, smoketestdisk) {
    FLAGS_zk_root_path = "/rtidb3" + ::openmldb::test::GenRand();
    brpc::Server tablet;
    ASSERT_TRUE(::openmldb::test::StartTablet("127.0.0.1:9530", &tablet));

    brpc::Server server;
    ASSERT_TRUE(::openmldb::test::StartNS("127.0.0.1:9631", "127.0.0.1:9530", &server));
    ::openmldb::client::NsClient client("127.0.0.1:9631", "");
    ASSERT_EQ(client.Init(), 0);

    TableInfo table_info;
    table_info.set_storage_mode(openmldb::common::kHDD);
    std::string name = "test" + ::openmldb::test::GenRand();
    table_info.set_name(name);
    ::openmldb::test::AddDefaultSchema(0, 0, ::openmldb::type::kAbsoluteTime, &table_info);
    std::string msg;
    ASSERT_TRUE(client.CreateTable(table_info, false, msg));

    openmldb::base::RemoveDirRecursive(FLAGS_hdd_root_path);
}


}  // namespace nameserver
}  // namespace openmldb

int main(int argc, char** argv) {
    FLAGS_zk_session_timeout = 100000;
    ::testing::InitGoogleTest(&argc, argv);
    ::openmldb::base::SetLogLevel(INFO);
    ::google::ParseCommandLineFlags(&argc, &argv, true);
    ::openmldb::test::TempPath tmp_path;
    FLAGS_db_root_path = tmp_path.GetTempPath();
    FLAGS_hdd_root_path = tmp_path.GetTempPath();
    return RUN_ALL_TESTS();
}

File: test/integration-test/openmldb-test-java/openmldb-http-test/src/main/java/com/_4paradigm/openmldb/http_test/common/StandaloneTest.java
/*
 * Copyright 2021 4Paradigm
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com._4paradigm.openmldb.http_test.common;


import com._4paradigm.openmldb.test_common.openmldb.OpenMLDBGlobalVar;
import com._4paradigm.qa.openmldb_deploy.bean.OpenMLDBDeployType;
import com._4paradigm.qa.openmldb_deploy.bean.OpenMLDBInfo;
import com._4paradigm.qa.openmldb_deploy.common.OpenMLDBDeploy;
import com.google.common.collect.Lists;
import lombok.extern.slf4j.Slf4j;
import org.apache.commons.lang3.StringUtils;
import org.testng.annotations.BeforeTest;
import org.testng.annotations.Optional;
import org.testng.annotations.Parameters;

@Slf4j
public class StandaloneTest extends BaseTest{

    @BeforeTest()
    @Parameters({"env", "version", "fedbPath"})
    public void beforeTest(@Optional("qa") String env, @Optional("main") String version, @Optional("") String fedbPath) throws Exception {
        RestfulGlobalVar.env = env;
        String caseEnv = System.getProperty("caseEnv");
        if (!StringUtils.isEmpty(caseEnv)) {
            RestfulGlobalVar.env = caseEnv;
        }
        log.info("openmldb global var env: {}", RestfulGlobalVar.env);
        if(env.equalsIgnoreCase("standalone")){
            OpenMLDBDeploy fedbDeploy = new OpenMLDBDeploy(version);
            fedbDeploy.setOpenMLDBPath(fedbPath);
            RestfulGlobalVar.mainInfo = fedbDeploy.deployStandalone();
        }else{
            OpenMLDBInfo openMLDBInfo = new OpenMLDBInfo();
            openMLDBInfo.setDeployType(OpenMLDBDeployType.STANDALONE);
            openMLDBInfo.setHost("172.24.4.55");
            openMLDBInfo.setPort(30013);
            openMLDBInfo.setNsNum(1);
            openMLDBInfo.setTabletNum(1);
            openMLDBInfo.setBasePath("/home/wangkaidong/fedb-auto-test/standalone");
            openMLDBInfo.setZk_cluster("172.24.4.55:30000");
            openMLDBInfo.setZk_root_path("/openmldb");
            openMLDBInfo.setNsEndpoints(Lists.newArrayList("172.24.4.55:30013"));
            openMLDBInfo.setTabletEndpoints(Lists.newArrayList("172.24.4.55:30014"));
            openMLDBInfo.setApiServerEndpoints(Lists.newArrayList("172.24.4.55:30015"));
            openMLDBInfo.setOpenMLDBPath("/home/wangkaidong/fedb-auto-test/standalone/openmldb-standalone/bin/openmldb");

            RestfulGlobalVar.mainInfo = openMLDBInfo;
        }
    }
}
